### PR TITLE
Added doubt and doubt answer implementation with schemas , dtos and completed endpoints and population

### PR DESCRIPTION
**Pull request by Devesh21700Kumar**
**Description**

Added doubt and doubt answer implementation with schemas , dtos and completed endpoints and population

1). doubt Schema added
2). doubt Answer Schema added
3). field populated in course schema

- [x] Code
- [x] New Feature

- [x] My code follows the code style of this project.
- [x] All new and existing tests passed.
- [x] My changes generate no new warnings
- [x] The title of my pull request is a short description of the requested changes.

*How Has This Been Tested?*
Terminal commands to test code:-
1). npm run lint
2). npm run lint:fix
3). nest start
4). npm run test

Also implemented deep population to populate answers of doubt schema in course as well
image

### CODE REVIEW

**Review by Devesh21700Kumar — COMMENTED** [review]
@Abhishek-kumar09 this is the doubt implementation using deep population

File: src/modules/doubt/schema/doubtAnswer.schema.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
+import * as mongoose from 'mongoose';
+
+export type DoubtAnswerDocument = DoubtAnswer & mongoose.Document;
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/schema/doubtAnswer.schema.ts:4]: **Dont Use `*`**

File: src/modules/course/schema/course.schema.ts [hunk 2]
```diff
@@ -69,6 +70,8 @@ export class Course {
   @Prop({ type: [{ type: SchemaTypes.Types.ObjectId, ref: 'Review' }] })
   reviews: Review[];
 
+  @Prop({ type: [{ type: SchemaTypes.Types.ObjectId, ref: 'Doubt' }] })
+  doubts: Doubt[];
```
Comment (Abhishek-kumar09) [inline src/modules/course/schema/course.schema.ts:74]: Add a line break

**Review by Abhishek-kumar09 — COMMENTED** [review]
Add Api operation in Doubt controller

File: src/modules/doubt/dto/update-doubt.dto.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  IsEnum,
+  IsNotEmpty,
+  IsOptional,
+  IsString,
+} from 'class-validator';
+import { TagType } from '../doubt-tag.enum';
+
+export class UpdateDoubtDto {
+  @IsNotEmpty()
+  @IsEnum(TagType)
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/dto/update-doubt.dto.ts:12]: Also Add is Array and check if it is working correctly

File: src/modules/course/schema/course.schema.ts (hunk 2 above)
Comment (Devesh21700Kumar) [reply src/modules/course/schema/course.schema.ts:74]: Sure

File: src/modules/doubt/dto/update-doubt.dto.ts (hunk 3 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/dto/update-doubt.dto.ts:12]: Ok

**Comment by Devesh21700Kumar** [comment]
@Abhishek-kumar09 changes made

File: src/modules/doubt/doubt.service.ts [hunk 4]
```diff
@@ -4,41 +4,213 @@ import {
...
+  async editDoubt(
+    courseId: Schema.Types.ObjectId,
+    doubtId: Schema.Types.ObjectId,
+    updateDoubtDto: UpdateDoubtDto,
+  ) {
+    try {
+      const course = await this.CourseModel.findById(courseId);
+      if (course) {
+        let updatedDoubt = null;
+        updatedDoubt = await this.DoubtModel.findByIdAndUpdate(
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/doubt.service.ts:57]: lean

File: src/modules/doubt/doubt.service.ts [hunk 5]
```diff
@@ -4,41 +4,213 @@ import {
...
+    courseId: Schema.Types.ObjectId,
+    doubtId: Schema.Types.ObjectId,
+  ): Promise<any> {
+    try {
+      const course = await this.CourseModel.findById(courseId);
+      if (course) {
+        let deletedDoubt = null;
+        deletedDoubt = await this.DoubtModel.findByIdAndRemove(
+          doubtId,
+        ).populate('answers');
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/doubt.service.ts:90]: lean

File: src/modules/doubt/doubt.service.ts [hunk 6]
```diff
@@ -4,41 +4,213 @@ import {
...
+    createDoubtDto: CreateDoubtDto,
+  ) {
     try {
-      return await this.DoubtModel.find().exec();
+      const course = await this.CourseModel.findById(courseId);
+      if (course) {
+        const newDoubt = await new this.DoubtModel(createDoubtDto).save();
+        course.doubts.push(newDoubt);
+        await course.save();
+        return newDoubt.populate('answers');
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/doubt.service.ts:36]: lean

File: src/modules/doubt/schema/doubtAnswer.schema.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+DoubtAnswerSchema.methods.toJSON = function () {
+  const doubtAnswerObject = this.toObject();
+  doubtAnswerObject.id = doubtAnswerObject._id;
+
+  delete doubtAnswerObject._id;
+  delete doubtAnswerObject.__v;
+  delete doubtAnswerObject['createdAt'];
+  delete doubtAnswerObject['updatedAt'];
+
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/schema/doubtAnswer.schema.ts:25]: remoive this

File: src/modules/doubt/schema/doubt.schema.ts [hunk 8]
```diff
@@ -1,26 +1,40 @@
...
+@Schema({ timestamps: true })
 export class Doubt {
   @Prop({ required: true })
-  tags: string[];
+  tags: TagType[];
 
-  @Prop({ type: SchemaType.Types.ObjectId, ref: 'User' })
-  asked_by: User;
+  @Prop({ required: true })
+  asked_by: string;
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/schema/doubt.schema.ts:15]: User

File: src/modules/doubt/doubt.service.ts (hunk 4 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/doubt.service.ts:57]: Ok

File: src/modules/doubt/doubt.service.ts (hunk 5 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/doubt.service.ts:90]: Ok

File: src/modules/doubt/schema/doubtAnswer.schema.ts (hunk 7 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/schema/doubtAnswer.schema.ts:25]: Ok

**Review by Devesh21700Kumar — COMMENTED** [review]
Fixed the asked_by and asnwered_by type by setting it to Object Id and added timestamps in doubtanswers

**Review by Devesh21700Kumar — COMMENTED** [review]
Fixed stuff and added getallDoubts and getDoubtbyId for doubtforums along with the course doubts section

**Review by Devesh21700Kumar — COMMENTED** [review]
@Abhishek-kumar09  I think that this is ready for merge as well

File: src/modules/course/schema/course.schema.ts [hunk 9]
```diff
@@ -4,6 +4,7 @@ import { Schedule } from './schedule.schema';
 import { TagType } from '../course-tag.enum';
 import { Review } from './review.schema';
 import { courseLevelType } from '../courseLevel.enum';
+import { Doubt } from 'modules/doubt/schema/doubt.schema';
```
Comment (Abhishek-kumar09) [inline src/modules/course/schema/course.schema.ts:7]: relative import

File: src/modules/doubt/doubt.service.ts [hunk 10]
```diff
@@ -4,41 +4,239 @@ import {
   NotFoundException,
 } from '@nestjs/common';
 import { InjectModel } from '@nestjs/mongoose';
-import { Model } from 'mongoose';
+import { Model, Schema } from 'mongoose';
 import { CreateDoubtDto } from './dto/create-doubt.dto';
 import { DoubtDocument as Doubt } from './schema/doubt.schema';
+import { Course } from 'modules/course/schema/course.schema';
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/doubt.service.ts:10]: wrong import

File: src/modules/doubt/doubt.service.ts [hunk 11]
```diff
@@ -4,41 +4,239 @@ import {
...
+        .populate('answers')
+        .lean();
 
       if (doubt) {
         return doubt;
       }
     } catch (e) {
       throw new InternalServerErrorException(e);
     }
+    throw new NotFoundException('Error');
```
Comment (Abhishek-kumar09) [inline src/modules/doubt/doubt.service.ts:48]: Provide verbose response

**Review by Abhishek-kumar09 — CHANGES_REQUESTED** [review]

File: src/modules/doubt/doubt.service.ts (hunk 10 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/doubt.service.ts]: Fixed

File: src/modules/course/schema/course.schema.ts (hunk 9 above)
Comment (Devesh21700Kumar) [reply src/modules/course/schema/course.schema.ts]: Fixed

File: src/modules/doubt/doubt.service.ts (hunk 11 above)
Comment (Devesh21700Kumar) [reply src/modules/doubt/doubt.service.ts]: Fixed

**Pull request merged**